twisted/twisted/plugins/tribler_plugin.py

```diff
@@ -0,0 +1,94 @@
+"""
+This twistd plugin enables to start Tribler headless using the twistd command.
+"""
+import logging
+import os
+import signal
+from time import sleep
+
+from twisted.application.service import MultiService, IServiceMaker
+from twisted.conch import manhole_tap
+from twisted.internet import reactor
+from twisted.plugin import IPlugin
+from twisted.python import usage
+from twisted.python.log import msg
+from zope.interface import implements
+
+from Tribler.Core.Session import Session
+from Tribler.Core.SessionConfig import SessionStartupConfig
+
+
+class Options(usage.Options):
+    optParameters = [
+        ["manhole", "m", 0, "Enable manhole telnet service listening at the specified port", int],
+    ]
+
+
+class TriblerServiceMaker(object):
+    implements(IServiceMaker, IPlugin)
+    tapname = "tribler"
+    description = "Tribler twistd plugin, starts Tribler as a twistd process"
+    options = Options
+
+    def __init__(self):
+        """
+        Initialize the variables of the TriblerServiceMaker and the logger.
+        """
+        self.session = None
+        self._stopping = False
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+    def start_tribler(self):
+        """
+        Main method to startup Tribler.
+        """
+
+        def signal_handler(sig, _):
+            msg("Received shut down signal %s" % sig)
+            if not self._stopping:
+                self._stopping = True
+                try:
+                    self.session.shutdown()
+                    msg("Tribler shut down")
+                except Exception, exc:
+                    msg("Got exception when stopping Tribler: %s" % exc)
+                reactor.stop()
+        signal.signal(signal.SIGINT, signal_handler)
+        signal.signal(signal.SIGTERM, signal_handler)
+
+        msg("Starting Tribler")
+
+        config = SessionStartupConfig()
+        config.set_http_api_enabled(True)
+        self.session = Session(config)
+        upgrader = self.session.prestart()
+        while not upgrader.is_done:
+            sleep(0.1)
+        self.session.start()
+
+        msg("Tribler started")
+
+    def makeService(self, options):
+        """
+        Construct a Tribler service.
+        """
+        tribler_service = MultiService()
+        tribler_service.setName("Tribler")
+
+        manhole_namespace = {}
+        if options["manhole"]:
+            port = options["manhole"]
+            manhole = manhole_tap.makeService({
+                'namespace': manhole_namespace,
+                'telnetPort': 'tcp:%d:interface=127.0.0.1' % port,
+                'sshPort': None,
+                'passwd': os.path.join(os.path.dirname(__file__), 'passwd'),
+            })
+            tribler_service.addService(manhole)
+
+        reactor.exitCode = 0
```

Comment:
I copied most of the file from the Dispersy twistd plugin and this line got copied over. I removed it, tested the shutdown procedure and it seems that the statement is indeed not necessary.